Apply loop contracts only if there exists some usage

qinheping opened the pull request:
In this PR, we detect the usage of loop contracts, and apply loop contracts with goto-instrument only if there exists some usage of loop contracts in the codegen unit.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval reviewed (commented):
Do we have any tests that check what is Kani's behavior when a loop contract is reachable but user did not enable it? If not, can you please add one.

qinheping commented:
@celinval I added two tests in this PR: one for harness without loop but loop contracts is enabled. One for harness with loop contracts, but loop contracts is not enabled. In both harness, Kani performs as no loop contracts present---unwind all loops.

We should apply loop contracts only when there exist some usages of loop contracts **and** the loop contracts is enabled by users.

I also fixed a bug in `hook.rs` of not correctly handling `loop_contracts_register_function` when loop contracts is not enabled.

celinval commented on kani-compiler/src/codegen_cprover_gotoc/overrides/hooks.rs:
```diff
@@ -567,25 +567,52 @@ impl GotocHook for LoopInvariantRegister {
-            loc,
-        )
+
+        gcx.has_loop_contracts = true;
```
Should this be inside the `if` block?

qinheping replied:
No, the hook is applied when there are calls to `register_loop_contracts` in MIR, which means there are loop contracts in the harness. The if block is about whether the loop contracts feature of Kani is enabled.

celinval reviewed (approved):
Thanks
